src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -612,8 +615,57 @@ public virtual ProtoCore.Type GetTypeHintForOutput(int index)
         {
              return ProtoCore.TypeSystem.BuildPrimitiveTypeObject(ProtoCore.PrimitiveType.kTypeVar);
         }
-        #endregion
 
+      
+        /// <summary>
+        /// Gets or sets a value indicating whether the node is frozen.
+        /// If the node is set to freeze by the user, then Freeze always true.
+        /// Otherwise, if any of the node has input node frozen, then Freeze is true.        
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if this node is frozen; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsFrozen
+        {
+            get
+            {
+                var gatheredInputFrozenNodes = new List<NodeModel>();
+                CheckIfAnyUpstreamNodeIsFrozen(this, gatheredInputFrozenNodes);
+
+                if (gatheredInputFrozenNodes.Any())
+                    return true;
+ 
+                return explictFrozen;
+            }
+            set
+            {
+                explictFrozen = value;                 
+                RaisePropertyChanged("IsFrozen");
```

Comment:
I don't think we need/want this property change notification here. When the user checks `Frozen` on the node in the view, the view model should raise a property change notification to ensure that the view does the right thing. But I don't think we need to do this on the model as well.